IllegalArgumentException while solving

When the symbol solver, tries to resolve `EntrySet<Object, Object>` from the code below it throws an IllegalArgumentException.



```java
class A {

    class EntrySetImpl implements EntrySet<Object, Object> {}

    class Bar {
        private class EntrySet {}
    }

    interface EntrySet<K, V> {}

}
```

I think the problem here, is when the symbol solver tries to find a class with the name `EntrySet` it finds first the class `A.Bar.EntrySet` rather than `A.EntrySet`.

<details>
<summary>Stack trace</summary>

```
Caused by: java.lang.IllegalArgumentException: expected either zero type arguments or has many as defined in the declaration (0). Found 2
	at com.github.javaparser.resolution.types.ResolvedReferenceType.<init>(ResolvedReferenceType.java:78)
	at com.github.javaparser.symbolsolver.model.typesystem.ReferenceTypeImpl.<init>(ReferenceTypeImpl.java:77)
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserClassDeclaration.toReferenceType(JavaParserClassDeclaration.java:481)
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserClassDeclaration.getAncestors(JavaParserClassDeclaration.java:354)
	at com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration.getAncestors(ResolvedReferenceTypeDeclaration.java:68)
	at com.github.javaparser.Issue.testCompilationUnitWithTwoClassesWithTheSameName(Issue.java:59)
```
</details>

<details>
<summary>Unit test</summary>

```java
@Test
void testCompilationUnitWithTwoClassesWithTheSameName() {
  // Setup symbol solver
  StaticJavaParser.getConfiguration()
		  .setSymbolResolver(
				  new JavaSymbolSolver(new ReflectionTypeSolver())
		  );
  // Setup source code
  String sourceCode = "class A {\n" +
		  "\n" +
		  "    class EntrySetImpl implements EntrySet<Object, Object> {}\n" +
		  "\n" +
		  "    class Bar {\n" +
		  "        private class EntrySet {}\n" +
		  "    }\n" +
		  "\n" +
		  "    interface EntrySet<K, V> {}\n" +
		  "\n" +
		  "}\n";
  CompilationUnit cu = StaticJavaParser.parse(sourceCode);

  // Resolve the EntrySetImpl class and try to get its ancestors
  ResolvedReferenceTypeDeclaration resolvedClass = cu.findFirst(
		  ClassOrInterfaceDeclaration.class,
		  c -> c.getNameAsString().equals("EntrySetImpl")
  ).get().resolve();
  try {
	  resolvedClass.getAncestors();
  } catch (IllegalArgumentException e) {
	  fail("Unable to resolve class EntrySet<Object, Object>", e);
  }
}
```
</details>

--- a/javaparser-core/src/main/java/com/github/javaparser/ast/Node.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/ast/Node.java
@@ -891,6 +891,19 @@ public abstract class Node implements Cloneable, HasParentNode<Node>, Visitable,
         walk(nodeType, found::add);
         return found;
     }
+    
+    /**
+     * Walks the AST with specified traversal order, returning all nodes of type "nodeType".
+     */
+    public <T extends Node> List<T> findAll(Class<T> nodeType, TreeTraversal traversal) {
+        final List<T> found = new ArrayList<>();
+        walk(traversal, node -> {
+            if (nodeType.isAssignableFrom(node.getClass())) {
+                found.add(nodeType.cast(node));
+            }
+        });
+        return found;
+    }
 
     /**
      * Walks the AST with pre-order traversal, returning all nodes of type "nodeType" that match the predicate.

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/ClassOrInterfaceDeclarationContext.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/ClassOrInterfaceDeclarationContext.java
@@ -26,6 +26,7 @@ import java.util.List;
 import java.util.Optional;
 
 import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.Node.TreeTraversal;
 import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
 import com.github.javaparser.ast.body.TypeDeclaration;
 import com.github.javaparser.ast.type.ClassOrInterfaceType;
@@ -127,7 +128,7 @@ public class ClassOrInterfaceDeclarationContext extends AbstractJavaParserContex
         }
         
         // otherwise search for local type
-        List<TypeDeclaration> localTypes = wrappedNode.findAll(TypeDeclaration.class);
+        List<TypeDeclaration> localTypes = wrappedNode.findAll(TypeDeclaration.class, TreeTraversal.BREADTHFIRST);
         for (TypeDeclaration<?> localType : localTypes) {
             if (localType.getName().getId().equals(name)) {
                 return SymbolReference.solved(JavaParserFacade.get(typeSolver)

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3087Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3087Test.java
@@ -11,7 +11,7 @@ import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclar
 import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
 
-public class Issue3088Test extends AbstractResolutionTest {
+public class Issue3087Test extends AbstractResolutionTest {
     
     @Test
     void testCompilationUnitWithTwoClassesWithTheSameName() {

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3087Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3087Test.java
@@ -1,0 +1,51 @@
+package com.github.javaparser.symbolsolver;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+
+public class Issue3088Test extends AbstractResolutionTest {
+    
+    @Test
+    void testCompilationUnitWithTwoClassesWithTheSameName() {
+      // Setup symbol solver
+      StaticJavaParser.getConfiguration()
+              .setSymbolResolver(
+                      new JavaSymbolSolver(new ReflectionTypeSolver())
+              );
+      // Setup source code
+      String sourceCode = "class A {\n" +
+              "\n" +
+              "    class EntrySetImpl implements EntrySet<Object, Object> {}\n" +
+              "\n" +
+              "    class Bar {\n" +
+              "        private class EntrySet {}\n" +
+              "    }\n" +
+              "\n" +
+              "    interface EntrySet<K, V> {}\n" +
+              "\n" +
+              "}\n";
+      CompilationUnit cu = StaticJavaParser.parse(sourceCode);
+
+      // Resolve the EntrySetImpl class and try to get its ancestors
+      ClassOrInterfaceDeclaration coid = cu.findFirst(
+              ClassOrInterfaceDeclaration.class,
+              c -> c.getNameAsString().equals("EntrySetImpl")
+      ).get();
+      
+      ResolvedReferenceTypeDeclaration resolvedClass = coid.resolve();
+      try {
+          resolvedClass.getAncestors();
+      } catch (IllegalArgumentException e) {
+          fail("Unable to resolve class EntrySet<Object, Object>", e);
+      }
+    }
+
+}

